ktor-server/ktor-server-core/jvm/src/io/ktor/application/newapi/KtorPlugin.kt

```diff
@@ -0,0 +1,655 @@
+ */
+public fun <A : Pipeline<*, ApplicationCall>> A.plugin(plugin: InterceptionsHolder): InterceptionsHolder {
+    return attributes[featureRegistryKey].getOrNull(plugin.key)
+        ?: throw MissingApplicationFeatureException(plugin.key)
+}
+
+/**
+ * [Execution] holds everything that is bound to the current plugin execution. Instance of [Execution] can be
+ * usually accessed from inside of any callback that defines feature behaviour (ex.: [KtorPlugin.onCall],
+ * [KtorPlugin.onReceive], [KtorPlugin.onResponse], etc.).
+ *
+ * [Execution] also provides useful methods, variables and mechanisms that cen be used in multiple advanced scenarios.
+ *
+ * **Please note that every method of the  [Execution] class should be considered an
+ * advanced functionality and requires a deep knowledge and understanding of the internals of Ktor.
+ * Please, read about [Pipeline] and [PipelinePhase] first.**
+ *
+ * @param SubjectT is a type of subject for the current feature that is being processed. See [subject] for more information.
+ *
+ * */
+public inline class Execution<SubjectT : Any>(private val context: PipelineContext<SubjectT, ApplicationCall>) {
+    /**
+     * Continues execution of the HTTP pipeline starting after the current plugin.
+     * Current plugin's code stops being executed from the moment of calling this function.
+     * */
+    public suspend fun proceed(): SubjectT = context.proceed()
+
+    /**
+     * Continues execution of the HTTP pipeline starting after the current plugin.
+     * At the same time, the [subject] of the execution will be changed.
+     * Current plugin's code stops being executed from the moment of calling this function.
+     * */
+    public suspend fun proceedWith(subectT: SubjectT): SubjectT = context.proceedWith(subectT)
+
+    /**
+     * Finishes execution of the current plugin as well as the current pipeline (see [Pipeline] for more information).
+     * */
+    public fun finish(): Unit = context.finish()
+
+    /**
+     * A subject of the current execution. A subject can be any object of some allowed type that is currently being
+     * processed.
+     *
+     * - For the stage of processing a call, there is no [subject] (i.e. [SubjectT] is [Unit])
+     * - For the stage of sending a response, the [subject] is an object that is currently being sent via HTTP protocol. ([SubjectT] can be [Any] type)
+     * - For the stage of receiving a request data, the [subject] is of type [ApplicationReceiveRequest]. Please, read [ApplicationReceiveRequest] javadoc for more information.
+     * */
+    public val subject: SubjectT get() = context.subject
+
+    /**
+     * A call that is being processed by the plugin in the current execution.
+     * */
+    public val call: ApplicationCall get() = context.call
+
+    // Useful methods
+
+    /**
+     * See [ApplicationEnvironment] for more details.
+     * */
+    public val environment: ApplicationEnvironment get() = context.application.environment
```

Comment:
Here and below can be extension properties, but I don't insist.

Reply:
Ok, no problem